pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+            numiter = 2
+            for dd in range(numiter):
+                msgs.info(f"Iterating on spatial translation - ITERATION #{dd+1}/{numiter}")
+                # Setup the WCS to use for all white light images
+                ref_idx = None  # Don't use an index - This is the default behaviour when a reference image is supplied
+                image_wcs, voxedge, reference_image = self.create_wcs(self.all_ra[ww], self.all_dec[ww], self.all_wave[ww],
+                                                                      self._dspat, wavediff, collapse=True)
+                if voxedge[2].size != 2:
+                    msgs.error("Spectral range for WCS is incorrect for white light image")
+
+                wl_imgs = generate_image_subpixel(image_wcs, self.all_ra[ww], self.all_dec[ww], self.all_wave[ww],
+                                                  self.all_sci[ww], self.all_ivar[ww], self.all_wghts[ww],
+                                                  self.all_spatpos[ww], self.all_specpos[ww], self.all_spatid[ww],
+                                                  self.all_tilts, self.all_slits, self.all_align, voxedge,
+                                                  all_idx=self.all_idx[ww],
+                                                  spec_subpixel=self.spec_subpixel, spat_subpixel=self.spat_subpixel)
+                if reference_image is None:
+                    # ref_idx will be the index of the cube with the highest S/N
+                    ref_idx = np.argmax(self.weights)
+                    reference_image = wl_imgs[:, :, ref_idx].copy()
+                    msgs.info("Calculating spatial translation of each cube relative to cube #{0:d})".format(ref_idx+1))
+                else:
+                    msgs.info("Calculating the spatial translation of each cube relative to user-defined 'reference_image'")
+
+                # Calculate the image offsets relative to the reference image
+                for ff in range(self.numfiles):
+                    # Calculate the shift
+                    ra_shift, dec_shift = calculate_image_phase(reference_image.copy(), wl_imgs[:, :, ff], maskval=0.0)
+                    # Convert pixel shift to degrees shift
+                    ra_shift *= self._dspat/cosdec
+                    dec_shift *= self._dspat
+                    msgs.info("Spatial shift of cube #{0:d}: RA, DEC (arcsec) = {1:+0.3f} E, {2:+0.3f} N".format(ff+1, ra_shift*3600.0, dec_shift*3600.0))
+                    # Apply the shift
+                    self.all_ra[self.all_idx == ff] += ra_shift
+                    self.all_dec[self.all_idx == ff] += dec_shift
+
+    def compute_weights(self):
+        """
+        Compute the relative weights to apply to pixels that are collected into the voxels of the output DataCubes
+        """
+        # Calculate the relative spectral weights of all pixels
+        if self.numfiles == 1:
```

Comment:
This should also probably be moved to core. If the user cannot understand what comes in and what goes out, the control flow is impossible to comprehend.

Reply:
I still think  that there is no reason to have this method in this class since it does not specifically depend on complicated objects or specific methods for this class. The only class method you are calling is create_wcs which itself wraps a core method. From what I see here,  you could easily make this function work with lists of standard python inputs i.e. lists and arrays.  That would involve a longer set of argument lists, but would also be more transparent and more generalizable. For example, suppose I want to compute weights for a set of cubes for  fiber IFU or a JWST IFU, but for whatever reason cannot perform all the operations in this class.

Reply:
This was a painful change, but I think it's all there now.